tests/src/common/test_case.rs

```diff
@@ -288,9 +291,192 @@ impl<Dispatcher, Treasury, Profit, Leaser, Lpp, Oracle, TimeAlarms>
+        })
+    }
+
+    pub fn execute<'r, T>(
+        &'r mut self,
+        sender: Addr,
+        contract_addr: Addr,
+        msg: &T,
+        send_funds: &[CwCoin],
+    ) -> anyhow::Result<WrappedResponse<'r, AppResponse>>
+    where
+        T: Debug + Serialize,
+    {
+        self.with_mock_app(|app: &mut MockApp| {
+            app.execute_contract(sender, contract_addr, msg, send_funds)
+        })
+    }
+
+    pub fn execute_raw<T>(
+        &mut self,
+        sender: Addr,
+        msg: T,
+    ) -> anyhow::Result<WrappedResponse<'_, AppResponse>>
+    where
+        T: Into<CosmosMsg>,
+    {
+        self.with_mock_app(|app: &mut MockApp| app.execute(sender, msg.into()))
+    }
+
+    pub fn sudo<'r, T, U>(
+        &'r mut self,
+        contract_addr: T,
+        msg: &U,
+    ) -> anyhow::Result<WrappedResponse<'r, AppResponse>>
+    where
+        T: Into<Addr>,
+        U: Serialize,
+    {
+        self.with_mock_app(|app: &mut MockApp| app.wasm_sudo(contract_addr, msg))
+    }
+
+    pub fn with_mock_app<'r, F, R>(&'r mut self, f: F) -> anyhow::Result<WrappedResponse<'r, R>>
+    where
+        F: FnOnce(&'r mut MockApp) -> anyhow::Result<R>,
+    {
+        self.message_receiver.assert_empty();
+
+        match f(&mut self.app) {
+            Ok(result) => Ok(WrappedResponse {
+                receiver: &mut self.message_receiver,
+                result,
+            }),
+            Err(error) => {
+                // On error no messages should be "sent out".
+                while self.message_receiver.try_iter().next().is_some() {}
+
+                Err(error)
+            }
+        }
+    }
+
+    #[must_use]
+    pub fn query(&self) -> QuerierWrapper<'_, Empty> {
+        self.message_receiver.assert_empty();
+
+        self.app.wrap()
+    }
+}
+
+#[must_use]
+#[derive(Debug)]
+pub struct WrappedResponse<'r, T> {
+    receiver: &'r mut CustomMessageReceiverExt,
+    result: T,
+}
+
+impl<'r> Iterator for WrappedResponse<'r, ()> {
+    type Item = CustomMsg;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.receiver.try_recv().ok()
+    }
+}
+
+impl<'r, T> WrappedResponse<'r, T> {
+    #[must_use]
+    pub fn receiver(&mut self) -> &mut CustomMessageReceiverExt {
+        self.receiver
+    }
+
+    #[must_use]
+    pub fn iter(&mut self) -> MpscIter<'_, CustomMsg> {
+        self.receiver.try_iter()
+    }
+
+    pub fn clear_result(self) -> WrappedResponse<'r, ()> {
+        WrappedResponse {
+            receiver: self.receiver,
+            result: (),
+        }
+    }
+
+    #[must_use]
+    pub fn unwrap_response(self) -> T {
+        self.receiver.assert_empty();
+
+        self.result
+    }
+}
+
+impl<'r> WrappedResponse<'r, ()> {
+    pub fn set_result<T>(self, result: T) -> WrappedResponse<'r, T> {
+        WrappedResponse {
+            receiver: self.receiver,
+            result,
+        }
+    }
```

Comment:
this smells ... so let's rethink the need for "WrappedResponse" ... leaving the naming aside

Reply:
Why do you think it does? It is only implemented when the response data is from the unit type, `()`.